hazelcast/src/main/java/com/hazelcast/query/impl/predicates/PartitionPredicateImpl.java

```diff
@@ -95,22 +122,30 @@ public int getClassId() {
 
     @Override
     public void writeData(ObjectDataOutput out) throws IOException {
-        out.writeObject(partitionKey);
+        if (out.getVersion().isLessThan(Versions.V5_2)) {
+            out.writeObject(getSinglePartitionKey());
+        } else {
+            out.writeObject(partitionKeys);
+        }
         out.writeObject(target);
     }
 
     @Override
     public void readData(ObjectDataInput in) throws IOException {
-        this.partitionKey = in.readObject();
+        if (in.getVersion().isLessThan(Versions.V5_2)) {
```

Comment:
I don't think this is also going to work. 

From the paths I see, the versions are only set to identified header in the enterprise edition and are simply skipped in the open-source implementation while reading. See https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L155-L158 

So, the version you will see here will always be the version of the _current codebase_, not the version that sent this data, if you are using open-source. Do you guys do anything special for such cases @ahmetmircik?

Reply:
One way to solve this without version checks would be introducing a new implementation(MultiPartitionPredicateImpl) and keeping this one, and giving it a unique class id so that the old readers cannot read it. And that would be fine I suppose, as we don't expect old servers to work with `multiPartitionPredicate` because it is introduced in 5.2

Reply:
Two implementations of the interface sounds ok to me

Reply:
How does this work in other operations that use this pattern already? e.g. [PutAllOperation](https://github.com/hazelcast/hazelcast/blob/14ef0dbefa4a6fa00a50a8873aeea02ec6c3f5e4/hazelcast/src/main/java/com/hazelcast/map/impl/operation/PutAllOperation.java#L216-L234) uses `getVersion()` to optionally read/write a boolean

Reply:
The other places are fine. They are part of an enterprise feature called Rolling Upgrade which lets the adjacent minor versions of the Hazelcast members form a cluster during upgrades. 

But, this class is re-used by the client and members. Clients serialize this class and send it as part of the requests. Imagine that you are an open-source Hazelcast user, you have 5.2 servers and 5.0 clients. Now, these checks won't work, and you will try to read a single partition key as a set of partition keys. Similar thing holds true for 5.2 clients and 5.0 servers.

So, in short, if the class can be serialized by both members and clients, we should not rely on these version checks.

Reply:
@mdumandag I've created a `MultiPartitionPredicate` class. I haven't pushed it yet, as it's taking a while to run through all the tests and make sure I've not missed something. 

Will hopefully have this change pushed tomorrow.

Reply:
I've pushed what I have in the meantime if you want to place some early feedback @mdumandag / @ahmetmircik